lib/internal/modules/run_main.js

```diff
@@ -43,12 +43,14 @@ function shouldUseESMLoader(mainPath) {
    */
   const userImports = getOptionValue('--import');
   if (userLoaders.length > 0 || userImports.length > 0) { return true; }
-  const { readPackageScope } = require('internal/modules/cjs/loader');
-  // Determine the module format of the main
+
+  // Determine the module format of the entry point.
   if (mainPath && StringPrototypeEndsWith(mainPath, '.mjs')) { return true; }
   if (!mainPath || StringPrototypeEndsWith(mainPath, '.cjs')) { return false; }
+
+  const { readPackageScope } = require('internal/modules/cjs/loader');
   const pkg = readPackageScope(mainPath);
-  return pkg && pkg.data.type === 'module';
+  return pkg.data?.type === 'module' || getOptionValue('--experimental-type') === 'module';
```

Comment:
Previously pkg was optional. Is this line correct?

Reply:
Surprisingly, yes. I discovered this when doing that JSDoc PR; the only potential values of `pkg` are an object or `false`, and `(false).property` is `undefined` rather than throwing.